File: src/data/portfolio.ts
import portfolioRaw from '../../data/portfolio.json?raw';
import aiChatImage480 from '../assets/AIchat-480.webp';
import aiChatImage960 from '../assets/AIchat-960.webp';
import bikeImage470 from '../assets/Bikeimg-470.webp';
import bikeImage940 from '../assets/Bikeimg-940.webp';
import loanImage480 from '../assets/LOS-480.webp';
import loanImage960 from '../assets/LOS-960.webp';
import portfolioImage480 from '../assets/Portfolioimg-480.webp';
import portfolioImage960 from '../assets/Portfolioimg-960.webp';
import profilePortrait426 from '../assets/Person-426.webp';
import profilePortrait768 from '../assets/Person-768.webp';
import resumeUrl from '../assets/Resume.pdf';
import weatherImage480 from '../assets/Weatherimg-480.webp';
import weatherImage960 from '../assets/Weatherimg-960.webp';

export interface SocialLink {
  id: 'github' | 'linkedin' | 'instagram' | string;
  label: string;
  url: string;
  source_level?: string;
}

export interface PortfolioProfile {
  id: string;
  full_name: string;
  display_name: string;
  headline: string;
  positioning: string;
  summary: string;
  location: {
    city: string;
    state: string;
    postal_code: string;
    country: string;
  };
  contact: {
    email: string;
    phone_display: string;
    phone_e164: string;
  };
  social_links: SocialLink[];
  portfolio_url: string;
}

export interface PortfolioExperience {
  id: string;
  employer: string;
  title: string;
  location: string;
  work_mode: string;
  start_date: string;
  end_date: string | null;
  period_label: string;
  status_note?: string;
  highlights: string[];
  technologies: string[];
  source: string;
}

export interface PortfolioEducation {
  id: string;
  institution: string;
  location: string;
  qualification: string;
  completion_year: number;
  score: string;
  needs_confirmation?: boolean;
  note?: string;
  source: string;
}

export interface PortfolioAchievement {
  id: string;
  title: string;
  description: string;
  source: string;
}

export interface PortfolioSkill {
  name: string;
  evidence: string[];
}

export interface PortfolioSkillCategory {
  id: string;
  label: string;
  skills: PortfolioSkill[];
}

export interface PortfolioProject {
  id: string;
  title: string;
  aliases?: string[];
  short_description: string;
  highlights: string[];
  technologies: string[];
  categories: string[];
  links: {
    github?: string;
    live?: string;
  };
  source: string;
  source_note?: string;
}

export interface PortfolioCertification {
  id: string;
  title: string;
  issuer: string;
  year: number | null;
  asset: string;
}

export interface PortfolioData {
  schema_version: string;
  last_verified_from_sources: string;
  source_policy: {
    primary: string;
    secondary: string[];
    rules: string[];
  };
  profile: PortfolioProfile;
  experience: PortfolioExperience[];
  education: PortfolioEducation[];
  achievements: PortfolioAchievement[];
  skill_categories: PortfolioSkillCategory[];
  projects: PortfolioProject[];
  certifications: PortfolioCertification[];
  known_unknowns: string[];
}

export interface ImageAsset {
  src: string;
  srcSet: string;
  sizes: string;
  width: number;
  height: number;
}

const parsedPortfolio = JSON.parse(portfolioRaw) as PortfolioData;

if (parsedPortfolio.schema_version !== '1.0.0') {
  throw new Error(`Unsupported portfolio schema: ${parsedPortfolio.schema_version}`);
}

export const portfolio = parsedPortfolio;

export type PortfolioSectionId =
  | 'home'
  | 'about'
  | 'skills'
  | 'projects'
  | 'experience'
  | 'certifications'
  | 'resume'
  | 'contact';

export const SECTION_LINKS: ReadonlyArray<{ id: PortfolioSectionId; label: string }> = [
  { id: 'home', label: 'Home' },
  { id: 'about', label: 'About' },
  { id: 'experience', label: 'Experience' },
  { id: 'skills', label: 'Skills' },
  { id: 'projects', label: 'Projects' },
  { id: 'certifications', label: 'Certifications' },
  { id: 'resume', label: 'Resume' },
  { id: 'contact', label: 'Contact' },
];

export const TOUR_SECTION_IDS: readonly PortfolioSectionId[] = [
  'home',
  'experience',
  'skills',
  'projects',
  'certifications',
  'resume',
  'contact',
];

export const PROJECT_CATEGORIES = [
  'All',
  ...Array.from(new Set(portfolio.projects.flatMap((project) => project.categories))),
];

export const RESUME_URL = resumeUrl;
export const PROFILE_PORTRAIT: ImageAsset = {
  src: profilePortrait426,
  srcSet: `${profilePortrait426} 426w, ${profilePortrait768} 768w`,
  sizes: '(min-width: 640px) 384px, calc(100vw - 72px)',
  width: 426,
  height: 539,
};

const PROJECT_IMAGE_SIZES =
  '(min-width: 1280px) 380px, (min-width: 768px) 32vw, 80vw';

const projectImages: Record<string, ImageAsset> = {
  'ai-chat-application': {
    src: aiChatImage480,
    srcSet: `${aiChatImage480} 480w, ${aiChatImage960} 960w`,
    sizes: PROJECT_IMAGE_SIZES,
    width: 480,
    height: 226,
  },
  'portfolio-website': {
    src: portfolioImage480,
    srcSet: `${portfolioImage480} 480w, ${portfolioImage960} 960w`,
    sizes: PROJECT_IMAGE_SIZES,
    width: 480,
    height: 236,
  },
  'weather-dashboard': {
    src: weatherImage480,
    srcSet: `${weatherImage480} 480w, ${weatherImage960} 960w`,
    sizes: PROJECT_IMAGE_SIZES,
    width: 480,
    height: 269,
  },
  'rideasy-bike-booking': {
    src: bikeImage470,
    srcSet: `${bikeImage470} 470w, ${bikeImage940} 940w`,
    sizes: PROJECT_IMAGE_SIZES,
    width: 470,
    height: 264,
  },
  'loan-onboarding-system': {
    src: loanImage480,
    srcSet: `${loanImage480} 480w, ${loanImage960} 960w`,
    sizes: PROJECT_IMAGE_SIZES,
    width: 480,
    height: 224,
  },
};

const certificationAssets: Readonly<Record<string, ImageAsset>> = {};

export const getProjectImage = (projectId: string): ImageAsset | undefined => projectImages[projectId];

export const getCertificationAsset = (assetPath: string): ImageAsset | undefined => certificationAssets[assetPath];

export const slugify = (value: string) =>
  value
    .trim()
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, '-')
    .replace(/^-+|-+$/g, '');

export const getSkillId = (name: string) => slugify(name);

File: backend/app/services/portfolio.py
from __future__ import annotations

import copy
import json
from collections.abc import Iterator
from dataclasses import dataclass
from pathlib import Path
from typing import Any, ClassVar

from backend.app.schemas import SourceType


@dataclass(frozen=True, slots=True)
class KnowledgeDocument:
    id: str
    kind: SourceType
    title: str
    text: str
    url: str | None = None


class PortfolioDataError(RuntimeError):
    pass


class PortfolioRepository:
    """Loads and exposes the audited portfolio knowledge source."""

    REQUIRED_KEYS: ClassVar[set[str]] = {
        "profile",
        "experience",
        "education",
        "achievements",
        "skill_categories",
        "projects",
        "certifications",
    }

    def __init__(self, data_path: Path) -> None:
        self.data_path = data_path
        self._data = self._load()
        self._documents = tuple(self._build_documents())

    def _load(self) -> dict[str, Any]:
        try:
            with self.data_path.open("r", encoding="utf-8") as handle:
                data = json.load(handle)
        except (OSError, json.JSONDecodeError) as exc:
            raise PortfolioDataError(f"Unable to load portfolio data: {exc}") from exc
        missing = sorted(self.REQUIRED_KEYS.difference(data))
        if missing:
            raise PortfolioDataError(f"Portfolio data is missing keys: {', '.join(missing)}")
        return data

    @property
    def raw(self) -> dict[str, Any]:
        return copy.deepcopy(self._data)

    @property
    def documents(self) -> tuple[KnowledgeDocument, ...]:
        return self._documents

    @property
    def projects(self) -> list[dict[str, Any]]:
        return copy.deepcopy(self._data["projects"])

    @property
    def skill_categories(self) -> list[dict[str, Any]]:
        return copy.deepcopy(self._data["skill_categories"])

    def all_skills(self) -> list[str]:
        names = {
            skill["name"]
            for category in self._data["skill_categories"]
            for skill in category["skills"]
        }
        return sorted(names, key=str.casefold)

    def profile_payload(self, resume_url: str) -> dict[str, Any]:
        return {
            "profile": copy.deepcopy(self._data["profile"]),
            "experience": copy.deepcopy(self._data["experience"]),
            "education": copy.deepcopy(self._data["education"]),
            "achievements": copy.deepcopy(self._data["achievements"]),
            "certifications": copy.deepcopy(self._data["certifications"]),
            "resume_url": resume_url,
        }

    def get_project(self, project_id: str) -> dict[str, Any] | None:
        return next(
            (copy.deepcopy(project) for project in self._data["projects"] if project["id"] == project_id),
            None,
        )

    def _build_documents(self) -> Iterator[KnowledgeDocument]:
        data = self._data
        profile = data["profile"]
        social_text = ", ".join(
            f"{item['label']}: {item['url']}" for item in profile["social_links"]
        )
        yield KnowledgeDocument(
            id=profile["id"],
            kind="profile",
            title=f"About {profile['display_name']}",
            text=(
                f"{profile['summary']} Headline: {profile['headline']}. "
                f"Location: {profile['location']['city']}, {profile['location']['state']}, India."
            ),
            url=profile["portfolio_url"],
        )
        yield KnowledgeDocument(
            id="contact",
            kind="contact",
            title="Contact and professional links",
            text=(
                f"Email: {profile['contact']['email']}. Phone: {profile['contact']['phone_display']}. "
                f"Portfolio: {profile['portfolio_url']}. {social_text}"
            ),
        )
        for item in data["experience"]:
            yield KnowledgeDocument(
                id=item["id"],
                kind="experience",
                title=f"{item['title']} at {item['employer']}",
                text=" ".join(
                    [
                        f"{item['title']} at {item['employer']}, {item['period_label']}, {item['location']} ({item['work_mode']}).",
                        *item["highlights"],
                        f"Technologies: {', '.join(item['technologies'])}.",
                        item.get("status_note", ""),
                    ]
                ).strip(),
            )
        for item in data["education"]:
            yield KnowledgeDocument(
                id=item["id"],
                kind="education",
                title=f"{item['qualification']} — {item['institution']}",
                text=" ".join(
                    part
                    for part in [
                        f"{item['qualification']} at {item['institution']}, {item['location']}, completed {item['completion_year']}, {item['score']}.",
                        item.get("note"),
                    ]
                    if part
                ),
            )
        for category in data["skill_categories"]:
            evidence = "; ".join(
                f"{skill['name']} ({', '.join(skill['evidence'])})" for skill in category["skills"]
            )
            yield KnowledgeDocument(
                id=category["id"],
                kind="skills",
                title=category["label"],
                text=f"Verified skills and evidence: {evidence}.",
            )
        for project in data["projects"]:
            links = ", ".join(f"{name}: {url}" for name, url in project["links"].items())
            aliases = ", ".join(project.get("aliases", []))
            yield KnowledgeDocument(
                id=project["id"],
                kind="project",
                title=project["title"],
                text=" ".join(
                    part
                    for part in [
                        project["short_description"],
                        *project["highlights"],
                        f"Technologies: {', '.join(project['technologies'])}.",
                        f"Categories: {', '.join(project['categories'])}.",
                        f"Also known as: {aliases}." if aliases else None,
                        f"Links: {links}." if links else None,
                        project.get("source_note"),
                    ]
                    if part
                ),
                url=project["links"].get("live") or project["links"].get("github"),
            )
        for item in data["certifications"]:
            yield KnowledgeDocument(
                id=item["id"],
                kind="certification",
                title=item["title"],
                text=(
                    f"{item['title']}, {item['issuer']}, {item['year']}. "
                    f"Verification status: {item['verification_status']}. {item['verification_note']}"
                ),
            )
        for item in data["achievements"]:
            yield KnowledgeDocument(
                id=item["id"],
                kind="achievement",
                title=item["title"],
                text=item["description"],
            )
